File: generate_payload.zig
const std = @import("std");
const dbus = @import("dbus_msg.zig");

const hello_payload_inner = blk: {
    @setEvalBranchQuota(10000);
    var b = dbus.Buf{};
    b.serialise(
        dbus.MESSAGE_TYPE_METHOD_CALL,
        1,
        .{
            .{ .code = dbus.FIELD_PATH, .sig = "o", .val = "/org/freedesktop/DBus" },
            .{ .code = dbus.FIELD_INTERFACE, .sig = "s", .val = "org.freedesktop.DBus" },
            .{ .code = dbus.FIELD_MEMBER, .sig = "s", .val = "Hello" },
            .{ .code = dbus.FIELD_DESTINATION, .sig = "s", .val = "org.freedesktop.DBus" },
        },
        "",
        .{},
    );
    const len = b.off;
    var result: [len]u8 = undefined;
    @memcpy(&result, b.get());
    break :blk result;
};

const notify_payload_inner = blk: {
    @setEvalBranchQuota(10000);
    var b = dbus.Buf{};
    b.serialise(
        dbus.MESSAGE_TYPE_METHOD_CALL,
        2,
        .{
            .{ .code = dbus.FIELD_PATH, .sig = "o", .val = "/org/freedesktop/Notifications" },
            .{ .code = dbus.FIELD_INTERFACE, .sig = "s", .val = "org.freedesktop.Notifications" },
            .{ .code = dbus.FIELD_MEMBER, .sig = "s", .val = "Notify" },
            .{ .code = dbus.FIELD_DESTINATION, .sig = "s", .val = "org.freedesktop.Notifications" },
            .{ .code = dbus.FIELD_SIGNATURE, .sig = "g", .val = "susssasa{sv}i" },
        },
        "susssasa{sv}i",
        .{
            "tiny-notify",
            @as(u32, 0),
            "",
            "CI Failed!",
            "Your build broke.",
            [_][]const u8{},
            [_]struct {
                key: []const u8,
                value: struct { []const u8, u32 },
            }{},
            @as(i32, -1),
        },
    );
    const len = b.off;
    var result: [len]u8 = undefined;
    @memcpy(&result, b.get());
    break :blk result;
};

pub fn buildHello() []const u8 {
    return &hello_payload_inner;
}

pub fn buildNotify() []const u8 {
    return &notify_payload_inner;
}

File: generate_payload.py
"""Generate D-Bus message payloads for Hello and Notify method calls.

Hello method: D-Bus Specification v0.43, 2024-10-29, §3.3.1 org.freedesktop.DBus.
https://dbus.freedesktop.org/doc/dbus-specification.html

Notify method signature "susssasa{sv}i": Desktop Notifications Specification,
v1.3, 2024-08-18, §2 Basic Design.
https://specifications.freedesktop.org/notification/latest/
"""

from dbus_msg import new_method_call, DBusAddress, Message
import logging


def build_hello_payload():
    addr = DBusAddress(
        object_path="/org/freedesktop/DBus",
        bus_name="org.freedesktop.DBus",
        interface="org.freedesktop.DBus",
    )
    msg = new_method_call(addr, "Hello")
    return msg.serialise(serial=1)


def build_notify_payload():
    addr = DBusAddress(
        object_path="/org/freedesktop/Notifications",
        bus_name="org.freedesktop.Notifications",
        interface="org.freedesktop.Notifications",
    )
    body = ("tiny-notify", 0, "", "CI Failed!", "Your build broke.", [], {}, -1)
    msg = new_method_call(addr, "Notify", "susssasa{sv}i", body)
    return msg.serialise(serial=2)


def validate_hello(data):
    h = Message.from_buffer(data)
    return (
        h.header.fields[1] == "/org/freedesktop/DBus"
        and h.header.fields[6] == "org.freedesktop.DBus"
    )


def validate_notify(data):
    n = Message.from_buffer(data)
    return (
        n.header.fields[1] == "/org/freedesktop/Notifications"
        and n.body == ("tiny-notify", 0, "", "CI Failed!", "Your build broke.", [], {}, -1)
    )


def main(out_dir):
    hello_pkt = build_hello_payload()
    notify_pkt = build_notify_payload()

    (out_dir / "hello.bin").write_bytes(hello_pkt)
    (out_dir / "notify.bin").write_bytes(notify_pkt)

    logging.info("Hello: %s bytes", len(hello_pkt))
    logging.info("Notify: %s bytes", len(notify_pkt))

    assert validate_hello(hello_pkt)
    assert validate_notify(notify_pkt)
    logging.info("Validation OK")


if __name__ == "__main__":
    def _script_io():
        from pathlib import Path

        logging.basicConfig(level=logging.INFO, format="%(message)s")
        main(out_dir=Path.cwd())
        return 0

    raise SystemExit(_script_io())
